File: examples/scala/dtnecho.scala
#!/usr/bin/env scala-cli

// to install scala-cli see https://scala-cli.virtuslab.org/

//> using scala "3.1.2"
//> using lib "com.softwaremill.sttp.client3::core::3.6.2"
//> using lib "com.softwaremill.sttp.client3::okhttp-backend::3.6.2"
//> using lib "com.github.plokhotnyuk.jsoniter-scala::jsoniter-scala-core::2.13.31"
//> using lib "com.github.plokhotnyuk.jsoniter-scala::jsoniter-scala-macros::2.13.31"

import com.github.plokhotnyuk.jsoniter_scala.core.{JsonValueCodec, readFromArray, writeToArray}
import com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker
import sttp.capabilities.WebSockets
import sttp.client3.*
import sttp.client3.okhttp.OkHttpSyncBackend
import sttp.model.Uri
import sttp.ws.{WebSocket, WebSocketFrame}

import java.nio.charset.StandardCharsets
import java.util.Base64

/** API base path used for http request */
val api: String =
  val ip   = "127.0.0.1"
  val port = 3000
  s"http://$ip:$port"

val backend: SttpBackend[Identity, WebSockets] = OkHttpSyncBackend()

/** get uri body as string, throwing on any errors */
def uget(uri: Uri): String = backend.send(basicRequest.get(uri).response(asStringAlways)).body

// what follows are the data type and codec definitions for receiving and sending bundles
case class ReceivedBundle(bid: String, src: String, dst: String, data: String):
  def payload: Array[Byte] = Base64.getDecoder.decode(data)
case class SendBundle(src: String, dst: String, data: String, delivery_notification: Boolean, lifetime: Long)
given JsonValueCodec[ReceivedBundle] = JsonCodecMaker.make
given JsonValueCodec[SendBundle]     = JsonCodecMaker.make

@main def run(): Unit =
  val local_node = uget(uri"$api/status/nodeid")
  println(s"Running echo service on $local_node")
  // Define service endpoint, "echo" for 'dtn' nodes and '7' for 'ipn' nodes
  val service =
    if local_node.startsWith("ipn")
    then "7"
    else "echo"
  // Prior to receiving anything register the local service endpoint
  val register = uget(uri"$api/register?$service")
  println(s"registration message: $register")

  // this uses synchronous websocket for demo purposes, which is a bad idea in case of concurrent messages
  val ws = backend.send(basicRequest.get(uri"$api/ws").response(asWebSocketAlwaysUnsafe)).body

  def echoConfirmation(): Unit = println(ws.receiveText())

  // select json communication
  ws.sendText("/json")
  echoConfirmation()

  // ask to receive messages on the the given path
  ws.sendText(s"/subscribe $service")
  echoConfirmation()

  while true do
    // read incoming bundles
    val msg = readFromArray[ReceivedBundle](ws.receiveBinary(true))
    println(s"echoing ${new String(msg.payload)}")

    // send them back to the receiver
    ws.sendBinary(writeToArray(SendBundle(
      // change src and dst
      src = msg.dst,
      dst = msg.src,
      lifetime = 3600 * 24 * 1000,
      delivery_notification = false,
      data = msg.data
    )))
    echoConfirmation()

File: examples/python/dtnecho.py
#!/usr/bin/env python3

# More or less minimal example of a dtn7 echo service using websockets
#
# Requirements:
# pip3 install websocket-client
# pip3 install cbor2

import urllib.request
import websocket
from cbor2 import dumps, loads
import sys

PORT = sys.argv[1] if len(sys.argv) > 1 else 3000

# Ready to receive data?
recv_data = False

# Get the local node ID
local_node = urllib.request.urlopen(f"http://127.0.0.1:{PORT}/status/nodeid").read().decode("utf-8")
print("Running echo service on " + local_node)

# Define service endpoint, "echo" for 'dtn' nodes and '7' for 'ipn' nodes
service = "echo"
if local_node.startswith('ipn'):
    service = 7

# Prior to receiving anything register the local service endpoint
register = urllib.request.urlopen(f"http://127.0.0.1:{PORT}/register?{service}").read()
print(register)

def on_open(ws):
    print("Connected")

    # first switch to 'data' mode
    # we can then receive decoded bundles, giving us direct access to the payload
    # default would be 'bundle' mode where we have to manually decode the complete bundle
    ws.send("/data")

def on_message(ws, message):
    global recv_data
    global service

    #print(recv_data, message)
    if not recv_data:
        if message == "200 tx mode: data":
            print("mode changed to `data`")
            # after the mode was set we subscribe to the echo service previously registered
            if local_node.startswith('ipn'):
                service = "{}.{}".format(local_node.split(".")[0], service)
            ws.send("/subscribe " + service)
        elif message == "200 subscribed":
            print("successfully subscribed")
            # after subscribing we are ready to receive bundles
            recv_data = True
    else:
        if message[0:3] == '200':
            # text messages starting with '200' indicate successful transmission
            print("sent reply")
        else:
            #hexstr = "".join(format(i, "02x") for i in message)
            #print("decoding: " + hexstr)

            # load CBOR message as dictionary
            data = loads(message)

            print("received new bundle: " + data['bid'])

            # construct the echo reply, swapping 'src' and 'dst'
            out = {
                'src': data['dst'],
                'dst': data['src'],
                'delivery_notification' : False,
                'lifetime' : 3600*24*1000,
                'data' : data['data']
            }
            #[print(key,':',value) for key, value in out.items()]

            # encode the response as a CBOR byte string
            out_cbor = dumps(out)

            #hexstr = "".join(format(i, "02x") for i in out_cbor)
            #print("response: " + hexstr)

            # send cbor encoded data as binary (opcode = 2)
            ws.send(out_cbor, opcode=2)

# Enable debug output from websocket engine
# websocket.enableTrace(True)

# Connect to default port of dtn7 running on the local machine
wsapp = websocket.WebSocketApp(f"ws://127.0.0.1:{PORT}/ws", on_message=on_message, on_open=on_open)
wsapp.run_forever()

